### PR TITLE
Extend subgraph utils to cover merging a node following a subgraph

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* **#52513 Extend subgraph utils to cover merging a node following a subgraph**
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Subgraph Utils previously only worked with merging a node into a subgraph if the node was before the subgraph; extend the logic for the case where the subgraph is first.

Differential Revision: [D26696697](https://our.internmc.facebook.com/intern/diff/D26696697)

### CODE REVIEW

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 1]
```diff
@@ -26,9 +26,11 @@ std::vector<c10::optional<const Use>> gatherLastUses(
 // Values which do not have uses or which do not have a last use
 // outside of the subgraph to be merged into we do not need to track.
 struct ValueMapper {
-  ValueMapper(Node* to_merge, AliasDb& db, size_t subgraph_num_outputs) {
+  ValueMapper(Node* to_merge, AliasDb& db, c10::optional<Node*> existing) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:29]: could we relate the comment below to `ValueMapper` c-tor's parameters?

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 2]
```diff
@@ -96,28 +103,29 @@ Node* executeSubgraphMergeAndUpdateAliasing(
...
+    end_it = nodeAfterMergeFrom->iterator();
+  } else {
+    end_it = nodeBeforeMergeFrom->reverseIterator();
+    it = nodeAfterMergeFrom->reverseIterator();
+  }
   ++it;
 
-  // defer destroying nodes until after all nodes have been merged,
-  // to make iterators easier to reason about
   std::vector<Node*> merged_nodes;
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:124]: looks like this isn't used anymore?

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 3]
```diff
@@ -96,28 +103,29 @@ Node* executeSubgraphMergeAndUpdateAliasing(
...
   while (it != end_it) {
     Node* node = *it;
     ++it;
-    merged_nodes.push_back(node);
-    mergeNodeIntoSubgraph(node, mergeTo, /*destroyNode*/ false);
-  }
-
-  for (Node* n : merged_nodes) {
-    n->destroy();
+    mergeNodeIntoSubgraph(node, mergeTo);
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:128]: ` /*destroyNode*/ false` do you know why this was an issue before?

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 4]
```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
...
-        [&](const Use& use) { return use.user->isAfter(subgraphNode); });
-
-    if (hasUsesOutsideSubgraph) {
-      auto newOutput = mergedNode->outputs()[i];
-      subgraph->registerOutput(newOutput);
-      auto groupOutput = subgraphNode->addOutput();
-      groupOutput->copyMetadata(oldOutput);
-      oldOutput->replaceAllUsesWith(groupOutput);
-    }
+    auto newOutput = mergedNode->outputs()[i];
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:300]: why did we hoist pruning unused outputs outside and after this loop?

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 5]
```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
...
+    // If n's outputs were inputs to `group`, remove them since we just merged
+    // n in.
+    //
+    // i.e.,
+    // x = f(w); group(x, y, z) becomes group(w, y, z).
+    // x, y, z = f(w); group(x, y, z) becomes group(w).
+    auto inputs = subgraphNode->inputs();
+    for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
+      auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
+      if (it != inputs.end()) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:288]: I understand you didn't change this code but do you know when we can have `it == inputs.end()`?

**Review by Krovatkin — COMMENTED** [review]
a few questions

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:128]: this could have been removed with the removal of the vmap API PR, I think it existed for that reason

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:29]: i'm not sure what you mean

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:300]: Because we want to do it after we destroy the node

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 5 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:288]: > If n's outputs were inputs to `group`, remove them since we just merged 

If any of `n`'s outputs weren't inputs to the subgraph.

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 6]
```diff
@@ -202,20 +212,30 @@ void mergeNodeIntoSubgraph(
 
   auto subgraph = getSubgraph(subgraphNode);
 
-  // Map from values in the surrounding graph to inputs in the subgraph
-  std::unordered_map<Value*, Value*> inputsMap;
+  // Map from values in the surrounding graph to values in the subgraph
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:215]: Minor: tbh, this comment is hard to follow. Maybe add an example or change "values" to "inputs/outputs".

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:29]: I'm not sure what @Krovatkin meant, but I'd find it helpful if we had a description of what each argument of the constructor is used for (especially the last one).

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 4 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:300]: Might be worth adding a comment saying exactly that :)

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me, but please make sure @Krovatkin doesn't have any concerns regarding that either!
